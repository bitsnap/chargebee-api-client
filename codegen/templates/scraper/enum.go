package scraper

import (
	"github.com/PuerkitoBio/goquery"
	. "github.com/bitsnap/go-util"
	"github.com/gocolly/colly/v2"
)

func ScrapeEnum(abbr string, page string) (map[string]string, error) {
	collector := colly.NewCollector(
		colly.AllowedDomains("apidocs.chargebee.com", "chargebee.com"),
		colly.CacheDir("./.cache"),
	)

	values := make(map[string]string)

	collector.OnHTML("div.cb-list-group", func(lists *colly.HTMLElement) {
		filterAbbr := func(_ int, elem *goquery.Selection) bool {
			return elem.Find("div.cb-list-item samp").First().Text() == abbr
		}

		elem := lists.DOM.Find("div.cb-list-action").FilterFunction(filterAbbr).First()
		snakeCaseValues := elem.Find("samp.enum").Map(func(i int, v *goquery.Selection) string {
			return v.Text()
		})

		for _, v := range snakeCaseValues {
			values[SnakeToCamel(v, true)] = v
		}
	})

	collector.OnError(func(response *colly.Response, err error) {
		Logger().Errorw(response.Request.URL.String(), "error", err)
	})

	err := collector.Visit(page)
	if err != nil {
		return nil, err
	}

	return values, nil
}
