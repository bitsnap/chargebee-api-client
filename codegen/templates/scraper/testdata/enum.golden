# Generated by goldga. DO NOT EDIT.
[snapshots]
"Enums scraping when traversing documentation pages should be able to find enum values" = '''
(map[string]string) (len=2) {
 (string) (len=17) "EachSpecifiedItem": (string) (len=19) "each_specified_item",
 (string) (len=13) "InvoiceAmount": (string) (len=14) "invoice_amount"
}
'''
